### PR TITLE
cleanup of reflection and macros

### PR DESCRIPTION
**Pull request by xeno-by**
I'm actually in the middle of it, but decided to give it a spin anyways, since the volume of the changes is quite big. review by @paulp, the very first commit is review by @VladUreche

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Macros.scala [hunk 1]
```diff
@@ -654,35 +621,16 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
...
-    var metaType = unwrapRet(macroImpl.tpe)
-
     // downgrade from metalevel-0 to metalevel-1
-    def inferRuntimeType(metaType: Type): Type = metaType match {
-      case TypeRef(pre, sym, args) if sym.name == tpnme.Expr && args.length == 1 =>
-        args.head
-      case _ =>
-        AnyClass.tpe
+    var runtimeType = macroImpl.tpe.finalResultType.dealias match {
+      case TypeRef(pre, sym, runtimeType :: Nil) if sym == ExprClass => runtimeType
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala]:
```
case TypeRef(_, ExprClass, runtimeType :: Nil) => runtimeType
```

**Comment by paulp** [comment]
It's getting harder to find stuff to complain about.  This looks like progress to me.  ("LGTM")

**Comment by xeno-by** [comment]
GitHub won't pick my changes. Probably because of the recent maintenance. I'll reopen the pull request.

**Comment by xeno-by** [comment]
PLS REBUILD ALL

**Comment by VladUreche** [comment]
LGTM for the first commit

**Pull request merged**